Surprising behavior with `typeof` in signatures

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->
<!--
Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker.

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the CONTRIBUTING guidelines: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ
-->

<!-- If you have a QUESTION:
   THIS IS NOT A FORUM FOR QUESTIONS.
   Ask questions at http://stackoverflow.com/questions/tagged/typescript
    or https://gitter.im/Microsoft/TypeScript
-->

<!-- If you have a SUGGESTION:
  Most suggestion reports are duplicates, please search extra hard before logging a new suggestion.
  See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md
-->

<!-- If you have a BUG:
  Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  2.9.0-dev.20180323

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 

**Code**

```ts
declare let foo: number;

// bar is of type string
// return type is string
declare function test(foo: string, bar: typeof foo): typeof foo;
// bar is of type string
// return type is number
declare function test({foo}: {foo: string}, bar: typeof foo): typeof foo;
```

**Expected behavior:**

`typeof foo` should reference the same declaration regardless of where it's used in the signature.

**Actual behavior:**

`typeof foo` in return type refers to the outer variable if `foo` comes from a destructured parameter. `typeof foo` inside the type annotation of another parameter refers to the parameter as expected.

This becomes even more visible if you remove the outer `declare let foo: number;` which will result in a compile error: `Cannot find name 'foo'.`

**Playground Link:** http://www.typescriptlang.org/play/#src=declare%20let%20foo%3A%20number%3B%0A%0Adeclare%20function%20test(foo%3A%20string%2C%20bar%3A%20typeof%20foo)%3A%20typeof%20foo%3B%0Adeclare%20function%20test(%7Bfoo%7D%3A%20%7Bfoo%3A%20string%7D%2C%20bar%3A%20typeof%20foo)%3A%20typeof%20foo%3B

**Related Issues:**
#22769 possibly related issue where the parameter initializer refers to the wrong variable.

In `declare function test({foo}: {foo: string}, bar: typeof foo): typeof foo;`, there is an error at the second `foo`: `src/a.ts(1,70): error TS2304: Cannot find name 'foo'.`. You might have been confused because of #22790 ?

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1234,7 +1234,7 @@ namespace ts {
                                     lastLocation.kind === SyntaxKind.Parameter ||
                                     (
                                         lastLocation === (<FunctionLikeDeclaration>location).type &&
-                                        result.valueDeclaration.kind === SyntaxKind.Parameter
+                                        !!findAncestor(result.valueDeclaration, isParameter)
                                     );
                             }
                         }

--- a/tests/baselines/reference/functionReturnTypeQuery.js
+++ b/tests/baselines/reference/functionReturnTypeQuery.js
@@ -1,0 +1,7 @@
+//// [functionReturnTypeQuery.ts]
+declare let foo: number;
+
+declare function test1(foo: string, bar: typeof foo): typeof foo;
+declare function test2({foo}: {foo: string}, bar: typeof foo): typeof foo;
+
+//// [functionReturnTypeQuery.js]

--- a/tests/baselines/reference/functionReturnTypeQuery.symbols
+++ b/tests/baselines/reference/functionReturnTypeQuery.symbols
@@ -1,0 +1,19 @@
+=== tests/cases/compiler/functionReturnTypeQuery.ts ===
+declare let foo: number;
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 0, 11))
+
+declare function test1(foo: string, bar: typeof foo): typeof foo;
+>test1 : Symbol(test1, Decl(functionReturnTypeQuery.ts, 0, 24))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 2, 23))
+>bar : Symbol(bar, Decl(functionReturnTypeQuery.ts, 2, 35))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 2, 23))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 2, 23))
+
+declare function test2({foo}: {foo: string}, bar: typeof foo): typeof foo;
+>test2 : Symbol(test2, Decl(functionReturnTypeQuery.ts, 2, 65))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 3, 24))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 3, 31))
+>bar : Symbol(bar, Decl(functionReturnTypeQuery.ts, 3, 44))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 3, 24))
+>foo : Symbol(foo, Decl(functionReturnTypeQuery.ts, 3, 24))
+

--- a/tests/baselines/reference/functionReturnTypeQuery.types
+++ b/tests/baselines/reference/functionReturnTypeQuery.types
@@ -1,0 +1,19 @@
+=== tests/cases/compiler/functionReturnTypeQuery.ts ===
+declare let foo: number;
+>foo : number
+
+declare function test1(foo: string, bar: typeof foo): typeof foo;
+>test1 : (foo: string, bar: string) => string
+>foo : string
+>bar : string
+>foo : string
+>foo : string
+
+declare function test2({foo}: {foo: string}, bar: typeof foo): typeof foo;
+>test2 : ({ foo }: { foo: string; }, bar: string) => string
+>foo : string
+>foo : string
+>bar : string
+>foo : string
+>foo : string
+

--- a/tests/baselines/reference/typeGuardFunctionErrors.symbols
+++ b/tests/baselines/reference/typeGuardFunctionErrors.symbols
@@ -312,6 +312,7 @@ function b5({a, b, p1}, p2, p3): p1 is A {
 >p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 127, 18))
 >p2 : Symbol(p2, Decl(typeGuardFunctionErrors.ts, 127, 23))
 >p3 : Symbol(p3, Decl(typeGuardFunctionErrors.ts, 127, 27))
+>p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 127, 18))
 >A : Symbol(A, Decl(typeGuardFunctionErrors.ts, 0, 0))
 
     return true;
@@ -324,6 +325,7 @@ function b6([a, b, p1], p2, p3): p1 is A {
 >p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 131, 18))
 >p2 : Symbol(p2, Decl(typeGuardFunctionErrors.ts, 131, 23))
 >p3 : Symbol(p3, Decl(typeGuardFunctionErrors.ts, 131, 27))
+>p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 131, 18))
 >A : Symbol(A, Decl(typeGuardFunctionErrors.ts, 0, 0))
 
     return true;
@@ -337,6 +339,7 @@ function b7({a, b, c: {p1}}, p2, p3): p1 is A {
 >p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 135, 23))
 >p2 : Symbol(p2, Decl(typeGuardFunctionErrors.ts, 135, 28))
 >p3 : Symbol(p3, Decl(typeGuardFunctionErrors.ts, 135, 32))
+>p1 : Symbol(p1, Decl(typeGuardFunctionErrors.ts, 135, 23))
 >A : Symbol(A, Decl(typeGuardFunctionErrors.ts, 0, 0))
 
     return true;

--- a/tests/baselines/reference/typeGuardFunctionErrors.types
+++ b/tests/baselines/reference/typeGuardFunctionErrors.types
@@ -356,7 +356,7 @@ function b5({a, b, p1}, p2, p3): p1 is A {
 >p1 : any
 >p2 : any
 >p3 : any
->p1 : No type information available!
+>p1 : any
 >A : A
 
     return true;
@@ -370,7 +370,7 @@ function b6([a, b, p1], p2, p3): p1 is A {
 >p1 : any
 >p2 : any
 >p3 : any
->p1 : No type information available!
+>p1 : any
 >A : A
 
     return true;
@@ -385,7 +385,7 @@ function b7({a, b, c: {p1}}, p2, p3): p1 is A {
 >p1 : any
 >p2 : any
 >p3 : any
->p1 : No type information available!
+>p1 : any
 >A : A
 
     return true;

--- a/tests/cases/compiler/functionReturnTypeQuery.ts
+++ b/tests/cases/compiler/functionReturnTypeQuery.ts
@@ -1,0 +1,4 @@
+declare let foo: number;
+
+declare function test1(foo: string, bar: typeof foo): typeof foo;
+declare function test2({foo}: {foo: string}, bar: typeof foo): typeof foo;

